Bug 1603269 - Alt. View for Intermittent Failures View for Screen Reader

yogmel opened the pull request:
## Description of issue
It is noted that the Graph in Intermittent Failures View is not accessible via Screen Reader. There are two tables:
- Failures Count per Push
- Failures Count vs Push Count

## Proposed Solution
As discussed, we decided to try on an alternate view as a table. The user can change the view from Graph to Table, via a toggle button. 

![Screenshot of Intermittent Failures view, showing its main screen, with two tables, one above the other.](https://user-images.githubusercontent.com/3901809/72580011-81711b00-38b9-11ea-822a-add1f94a8e85.png)

## Testing
NVDA navigation to test if the table was SR accessible. Also compared with the graph and console log of data, to check if data in the table was being mirrored.

yogmel commented:
Some comments:
- I've decided to go with **React Table** library instead of Reactstrap, because of its **pagination** features. The idea was that the table would have a similar height of the graph. I would suggest a change of _style_ to differentiate from the table down below.

- Also, I started with IFV, because the graph was simpler, compared to Perfherder's Graph. I thought it was a good starting point for me.

MarcoZehe commented:
👍 for this approach.

sarah-clements commented on ui/intermittent-failures/GraphAlternateView.jsx:
```diff
@@ -0,0 +1,87 @@
+  };
+
+  const convertDateToString = date =>
+    date.toLocaleDateString('en-US', {
```
The dates should be in UTC.

klahnakoski replied:
Do we have a time->utc somewhere already in this project?

sarah-clements replied:
It's listed in the page title, so yes, it is being used elsewhere.

sarah-clements replied:
@yogmel The `moment` library is used in Intermittent Failures View to format dates and I think it's simpler to format in UTC time that way.

sarah-clements commented on ui/intermittent-failures/GraphAlternateView.jsx:
```diff
@@ -0,0 +1,87 @@
+/* eslint-disable array-callback-return */
```
In general we try to disable the rule unless its poorly written. I'm curious where you hit this rule.

sarah-clements commented on ui/intermittent-failures/GraphAlternateView.jsx:
```diff
@@ -0,0 +1,87 @@
+
+  graphData.map((item, index) => {
+    if (colNum === 1) {
+      item.data.map(itemOne => {
```
Since you're not returning anything in map, you can use a `for` loop instead.

klahnakoski replied:
...or use `forEach()`

sarah-clements commented on ui/intermittent-failures/GraphAlternateView.jsx:
```diff
@@ -0,0 +1,87 @@
+import React from 'react';
+import ReactTable from 'react-table';
+
+const GraphAlternateView = ({ graphData, colNum }) => {
```
Since `colNum` is only designed to take one or two columns, perhaps this could be changed so it defaults to 1 and there's only two if someone uses the prop. Otherwise if someone enters a number greater than 2 it'll fail silently.

sarah-clements commented on ui/intermittent-failures/GraphAlternateView.jsx:
```diff
@@ -0,0 +1,87 @@
+          dataPoint.x = convertDateToString(itemTwo.x);
+          dataPoint.y = itemTwo.y;
+
+          alternateGraph.data.push(dataPoint);
```
Some of these duplicate items you can move to function to execute in a `for` loop.

sarah-clements commented on ui/intermittent-failures/GraphAlternateView.jsx:
```diff
@@ -0,0 +1,87 @@
+  );
+};
+
+export default GraphAlternateView;
```
This function should have proptypes.

sarah-clements commented:
> Some comments:
> 
>     * I've decided to go with **React Table** library instead of Reactstrap, because of its **pagination** features. The idea was that the table would have a similar height of the graph. I would suggest a change of _style_ to differentiate from the table down below.
> 
Sure, how about choosing a color from each of the graphs to correspond with the table versions (so for the first graph/table, use a lighter shade of the orange for the highlighted rows)? Open to other ideas. Please also include titles for the tables that correspond with the graphs.

>     * Also, I started with IFV, because the graph was simpler, compared to Perfherder's Graph. I thought it was a good starting point for me.

Always a good call :)

Looks great so far! One thing I'm wondering is if the `show table view` button gives enough context for screenreader users. Perhaps a description or aria-specific label to explain that it'll show an accessible view of the two types of graphs would be helpful.

MarcoZehe commented:
> Looks great so far! One thing I'm wondering is if the `show table view` button gives enough context for screenreader users. Perhaps a description or aria-specific label to explain that it'll show an accessible view of the two types of graphs (including their names) would be helpful.

On one hand, I believe someone technical enough to actually deal with Treeherder/Perfherder would figure out that something labeled a table view is accessible, especially if at the same time recognizing that someone went to lengths to offer an alternative view at all. On the other hand, the more verbose button label makes this explicit. But then, extra button labels for screen readers are only best practice in exceptional cases, and I am not sure if, given the first part of my comment, this is such an exceptional case. Thoughts, @jcsteh?

jcsteh commented:
I don't think we need a screen reader specific description; I agree this
isn't one of those exceptional cases where we should do that. That said, I
wonder whether we should rename this to "Switch to table view" instead?
First, assuming this hasn't changed, the description of this PR suggests it
is a toggle. Second, for some reason, this seems clearer to me that this is
an alternate view that might help me. However, that might just be a "me"
thing.

klahnakoski commented on ui/intermittent-failures/GraphAlternateView.jsx:
```diff
@@ -0,0 +1,87 @@
+      Header: 'Push Count',
+      accessor: 'z',
+    },
+  ];
```
Please use better names than `x`, `y`, and `z`.  I suggest `date`, `failureCount` and `pushCount`.

klahnakoski commented on ui/intermittent-failures/GraphAlternateView.jsx:
```diff
@@ -0,0 +1,87 @@
+    });
+
+  graphData.map((item, index) => {
+    if (colNum === 1) {
```
move this check outside the `map()`

klahnakoski commented on ui/intermittent-failures/GraphAlternateView.jsx:
```diff
@@ -0,0 +1,87 @@
+        const dataPoint = {};
+
+        if (index > 0) {
+          alternateGraph.data[indexTwo].z = itemTwo.y;
```
I do not understand this logic:  This copy from failure count (`y`) to the push count (`z`) is not done on the first iteration?

yogmel commented:
Thank you, everyone, for your feedback. I have made a lot of changes, so I will write some overall notes:

![Intermittent Test Failures screenshot - from top to bottom, there is the title, the date interval and the number of bugs per push. Down below, there is a table whose title is "Failure Count Per Push". The table has pagination buttons and shows up to 10 rows. The first column is for Date and the second for Failure count per push. The last column has an orange background color. Down below, there are more buttons: from left to right: "Show graph View", "more tables", start to end date picker input and "update"](https://user-images.githubusercontent.com/3901809/73209837-d4539980-4127-11ea-982e-8412212cd1e8.png)
![Another screenshot - now showing the "Failure Count vs. Push Count" table. The table looks like the first one, with the difference that there are 3 rows - Date, Failure Count and Push Count. The second column has a blue background color and the third, a green one.](https://user-images.githubusercontent.com/3901809/73209908-f220fe80-4127-11ea-91ae-6f0860fe8410.png)


- Styles were added to both tables in `intermittent-failures.css`;

- Changed loops and overall logic for creating tables;

- Inserted props for custom `title` and `className`

- I have moved the date format to `intermittent-failures/helpers.js`. I needed to keep the `x` and `y` properties, due to VictoryChart. This data object feeds both the table and the graph.

sarah-clements commented:
I think the colors for only certain columns looks too visually "busy". Can you simplify it by making the light gray table rows (entire row only) light orange and white. I think doing that for both tables is sufficient because there will be titles for each table and this is only to differentiate it from the tables beneath.

For the titles, please add padding beneath it so it's not so close to the table. I think also making it the same font size/family as the subheader above it ( that shows the dates) would look good.

yogmel commented:
@sarah-clements sure, here are the screenshots for the changes:

![Intermittent Test Failures screenshot - from top to bottom, there is the title, the date interval and the number of bugs per push. Down below, there is a table whose title is "Failure Count Per Push". The first column is for Date and the second for Failure count per push. The row colors are white and light orange.](https://user-images.githubusercontent.com/3901809/73364658-b2722800-4289-11ea-9095-959bc20dd8f6.png)
!["Failure Count vs. Push Count" table. The table looks like the first one, with the difference that there are 3 rows - Date, Failure Count and Push Count. The rows are white and light orange.](https://user-images.githubusercontent.com/3901809/73364697-c28a0780-4289-11ea-961a-fc9d56004d6f.png)

sarah-clements commented on ui/css/intermittent-failures.css:
```diff
@@ -35,6 +35,56 @@ h1,
+  background-color: rgba(221, 102, 2, 0.5);
+}
+
+/* Failure Count vs Push Count Table */
```
Are all of these styles necessary with the new table row color change?

yogmel replied:
noo! I didn't send a push yet. I've sent the screenshot for approval. I will send a new push now.

sarah-clements commented:
That looks better thanks. Is this ready for another review?

yogmel commented:
> That looks better thanks. Is this ready for another review?

Yes, it is now :)

sarah-clements commented on ui/intermittent-failures/GraphAlternateView.jsx:
```diff
@@ -0,0 +1,89 @@
+import PropTypes from 'prop-types';
+
+const GraphAlternateView = ({ className, graphData, colNum, title }) => {
+  const columnsOne = [
```
These variables are a little confusing. Technically this is two columns and `columnsTwo` is three columns total... can you rename them to `twoColumns` and `threeColumns` respectively?

sarah-clements commented on ui/intermittent-failures/GraphAlternateView.jsx:
```diff
@@ -0,0 +1,89 @@
+    } else {
+      alternateGraph.column = columnsTwo;
+
+      item.data.forEach((itemTwo, indexTwo) => {
```
I think it's confusing if you name these `itemTwo` and `indexTwo`... just item and index should suffice.

sarah-clements commented on ui/intermittent-failures/GraphAlternateView.jsx:
```diff
@@ -0,0 +1,89 @@
+          ...alternateGraph.data[indexTwo],
+          ...(failureCount && { failureCount }),
+          ...(pushCount && { pushCount }),
+        };
```
Why can't this data structure look the same as line 45?         
```
alternateGraph.data.push({
 date,
 failureCount,
 pushCount
});
```

sarah-clements commented on ui/intermittent-failures/GraphAlternateView.jsx:
```diff
@@ -0,0 +1,89 @@
+};
+
+GraphAlternateView.defaultProps = {
+  colNum: 1,
```
I think a two column default makes sense.

yogmel commented:
> Why can't this data structure look the same as line 45?

**TL;DR**
I needed to find a way to insert the property value **only if it already existed** or if it was different than `undefined`, that is why I used this uncommon notation: `(pushCount && { pushCount })`

-------

Long answer:

@sarah-clements that is because I have to iterate two arrays with objects and return only one array with an object with another structure.

`[ { date: date1, failureCount: number }, (...) ]`

\+

`[ { date: date1, pushCount: number }, (...) ]`

=

`[ { date: date1, failureCount: number, pushCount: number }, (...) ]`

----------

On the console, the two arrays look like this:

![image](https://user-images.githubusercontent.com/3901809/73387113-cbd99b00-42ae-11ea-887d-71ba9c35aff4.png)

----------

I tried creating an empty array `[]` and then pushing an item each iteration. However, that created an array **double** the size.

I also tried the approach you suggested, but the problem is that on the second item iteration, as there is one property mission, its value is `undefined` and it is overwritten.

For example, the first array iteration, the array would look like: 
```javascript
[
  {
     date: "Jan 22",
     failureCount: 996
     pushCount: undefined // the first array doesn't have a pushCount property
  }
]
```

On the second array iteration, the array would look like: 
```javascript
[
  {
     date: "Jan 22",
     failureCount: undefined // the first array doesn't have a failureCount property
     pushCount: 139
  }
]
```

On the console, the resulting array would look like this:
![image](https://user-images.githubusercontent.com/3901809/73388357-d5640280-42b0-11ea-93f8-e666cc4025bf.png)

----------

Therefore, I needed to find a way to insert the property value **only if it already existed**.

So, after some research, I found [this article (item 8)](https://blog.bitsrc.io/6-tricks-with-resting-and-spreading-javascript-objects-68d585bdc83#0856) with this code snippet.

The solution I came up with, with comments:

```javascript
alternateGraph.data[index] = {
  date, // get the dates from the first iteration
  ...alternateGraph.data[index], // get all the properties from previous iteration
  ...(failureCount && { failureCount }), // insert new property IF failureCount is not undefined
  ...(pushCount && { pushCount }), // insert new property IF pushCount is not undefined
};
```

-----------------

Sorry for the long text. I find it hard to explain this concept. Sorry if I overexplained. 
If I was not clear, we talk at another time. Whatever you prefer.

sarah-clements commented:
Ah, I see what you're trying to do. You can use lodash's [ zipWith](https://lodash.com/docs/4.17.15#zipWith) to iterate through two arrays at once to create one array (and specify how you want that new array to be structured). This will greatly simply that bit of code. We use that library already so you just need to import that method into the file it's being used in.

A tip for the next time you write a piece of code that might not be easy to understand at a glance is to add a brief comment explaining why/what it's for. Your future self (and other developers) will thank you :)

yogmel commented:
@sarah-clements it took me lots of time to figure out, but I managed to use `zipWith`. Probably there is a better way to use it.

I also found a code refactor, so I pushed that in a previous commit.

sarah-clements commented on ui/intermittent-failures/GraphAlternateView.jsx:
```diff
@@ -39,20 +40,20 @@ const GraphAlternateView = ({ className, graphData, colNum, title }) => {
+      graphData[0].data,
+      graphData[1].data,
+      (a, b) => {
+        const { date, failureCount } = a;
```
Do we know for certain these two arrays are in order by date? It might be a good precaution to sort each by date first before trying to match them in this way.

yogmel replied:
I sorted the data in the `helpers.js` file. I thought it would be better to deliver the data ready to the component to render.

sarah-clements replied:
Actually, I think this isn't necessary. In the `calculateMetrics` function, the objects containing the date are all inserted at the same time and nothing is changing that order so it should be fine without this.

yogmel commented:
Thank you :)
